File: csharp/src/impl/ClassDescriptor.cs
namespace Volante.Impl
{
    using System;
    using System.Collections;
    using System.Reflection;
    using System.Runtime.InteropServices;
    using System.Diagnostics;
    using System.Text;
    using Volante;

    public sealed class ClassDescriptor : Persistent
    {
        internal ClassDescriptor next;
        internal String name;
        internal FieldDescriptor[] allFields;
        internal bool hasReferences;
        internal static Module lastModule;

        public class FieldDescriptor : Persistent
        {
            internal String fieldName;
            internal String className;
            internal FieldType type;
            internal ClassDescriptor valueDesc;
            [NonSerialized()]
            internal FieldInfo field;
            [NonSerialized()]
            internal bool recursiveLoading;
            [NonSerialized()]
            internal MethodInfo constructor;

            public bool equals(FieldDescriptor fd)
            {
                return fieldName.Equals(fd.fieldName)
                    && className.Equals(fd.className)
                    && valueDesc == fd.valueDesc
                    && type == fd.type;
            }
        }
        [NonSerialized()]
        internal Type cls;
        [NonSerialized()]
        internal bool hasSubclasses;
        [NonSerialized()]
        internal ConstructorInfo defaultConstructor;
        [NonSerialized()]
        internal bool resolved;
        [NonSerialized()]
        internal GeneratedSerializer serializer;

        public enum FieldType
        {
            tpBoolean,
            tpByte,
            tpSByte,
            tpShort,
            tpUShort,
            tpChar,
            tpEnum,
            tpInt,
            tpUInt,
            tpLong,
            tpULong,
            tpFloat,
            tpDouble,
            tpString,
            tpDate,
            tpObject,
            tpOid,
            tpValue,
            tpRaw,
            tpGuid,
            tpDecimal,
            tpLink,
            tpArrayOfBoolean,
            tpArrayOfByte,
            tpArrayOfSByte,
            tpArrayOfShort,
            tpArrayOfUShort,
            tpArrayOfChar,
            tpArrayOfEnum,
            tpArrayOfInt,
            tpArrayOfUInt,
            tpArrayOfLong,
            tpArrayOfULong,
            tpArrayOfFloat,
            tpArrayOfDouble,
            tpArrayOfString,
            tpArrayOfDate,
            tpArrayOfObject,
            tpArrayOfOid,
            tpArrayOfValue,
            tpArrayOfRaw,
            tpArrayOfGuid,
            tpArrayOfDecimal,
            tpLast
        }

        internal static int[] Sizeof = new int[] 
        {
            1, // tpBoolean,
            1, // tpByte,
            1, // tpSByte,
            2, // tpShort, 
            2, // tpUShort,
            2, // tpChar,
            4, // tpEnum,
            4, // tpInt,
            4, // tpUInt,
            8, // tpLong,
            8, // tpULong,
            4, // tpFloat,
            8, // tpDouble,
            0, // tpString,
            8, // tpDate,
            4, // tpObject,
            4, // tpOid,
            0, // tpValue,
            0, // tpRaw,
            16,// tpGuid,
            16,// tpDecimal,
            0, // tpLink,
            0, // tpArrayOfBoolean,
            0, // tpArrayOfByte,
            0, // tpArrayOfSByte,
            0, // tpArrayOfShort, 
            0, // tpArrayOfUShort,
            0, // tpArrayOfChar,
            0, // tpArrayOfEnum,
            0, // tpArrayOfInt,
            0, // tpArrayOfUInt,
            0, // tpArrayOfLong,
            0, // tpArrayOfULong,
            0, // tpArrayOfFloat,
            0, // tpArrayOfDouble,
            0, // tpArrayOfString,
            0, // tpArrayOfDate,
            0, // tpArrayOfObject,
            0, // tpArrayOfOid,
            0, // tpArrayOfValue,
            0, // tpArrayOfRaw,
            0, // tpArrayOfGuid,
            0  // tpArrayOfDecimal,
        };

        internal static Type[] defaultConstructorProfile = new Type[0];
        internal static object[] noArgs = new object[0];

#if CF
        static internal object parseEnum(Type type, String value) 
        {
            foreach (FieldInfo fi in type.GetFields()) 
            {
                if (fi.IsLiteral && fi.Name.Equals(value)) 
                {
                    return fi.GetValue(null);
                }
            }
            throw new ArgumentException(value);
        }
#endif

        public bool equals(ClassDescriptor cd)
        {
            if (cd == null || allFields.Length != cd.allFields.Length)
            {
                return false;
            }
            for (int i = 0; i < allFields.Length; i++)
            {
                if (!allFields[i].equals(cd.allFields[i]))
                {
                    return false;
                }
            }
            return true;
        }

        internal Object newInstance()
        {
            try
            {
                return defaultConstructor.Invoke(noArgs);
            }
            catch (System.Exception x)
            {
                throw new DatabaseException(DatabaseException.ErrorCode.CONSTRUCTOR_FAILURE, cls, x);
            }
        }

#if CF
        internal void generateSerializer() {}
#else
        private static CodeGenerator serializerGenerator = CodeGenerator.Instance;

        internal void generateSerializer()
        {
            if (!cls.IsPublic || defaultConstructor == null || !defaultConstructor.IsPublic)
                return;

            FieldDescriptor[] flds = allFields;
            for (int i = 0, n = flds.Length; i < n; i++)
            {
                FieldDescriptor fd = flds[i];
                switch (fd.type)
                {
                    case FieldType.tpValue:
                    case FieldType.tpArrayOfValue:
                    case FieldType.tpArrayOfObject:
                    case FieldType.tpArrayOfEnum:
                    case FieldType.tpArrayOfRaw:
                    case FieldType.tpLink:
                    case FieldType.tpArrayOfOid:
                        return;
                    default:
                        break;
                }
                FieldInfo f = flds[i].field;
                if (f == null || !f.IsPublic)
                    return;
            }
            serializer = serializerGenerator.Generate(this);
        }

        static private bool isObjectProperty(Type cls, FieldInfo f)
        {
            return typeof(PersistentWrapper).IsAssignableFrom(cls) && f.Name.StartsWith("r_");
        }
#endif

        MethodInfo GetConstructor(FieldInfo f, string name)
        {
            MethodInfo mi = typeof(DatabaseImpl).GetMethod(name, BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.DeclaredOnly);
            //return mi.BindGenericParameters(f.FieldType.GetGenericArguments());
            //TODO: verify it's MakeGenericMethod
            return mi.MakeGenericMethod(f.FieldType.GetGenericArguments());
        }

        internal static String getTypeName(Type t)
        {
            if (t.IsGenericType)
            {
                Type[] genericArgs = t.GetGenericArguments();
                t = t.GetGenericTypeDefinition();
                StringBuilder buf = new StringBuilder(t.FullName);
                buf.Append('=');
                char sep = '[';
                for (int j = 0; j < genericArgs.Length; j++)
                {
                    buf.Append(sep);
                    sep = ',';
                    buf.Append(getTypeName(genericArgs[j]));
                }
                buf.Append(']');
                return buf.ToString();
            }
            return t.FullName;
        }

        static bool isVolanteInternalType(Type t)
        {
            return t.Namespace == typeof(IPersistent).Namespace
                && t != typeof(IPersistent) && t != typeof(PersistentContext) && t != typeof(Persistent);
        }

        internal void buildFieldList(DatabaseImpl db, System.Type cls, ArrayList list)
        {
            System.Type superclass = cls.BaseType;
            if (superclass != null && superclass != typeof(MarshalByRefObject))
            {
                buildFieldList(db, superclass, list);
            }
            System.Reflection.FieldInfo[] flds = cls.GetFields(BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.DeclaredOnly);
#if !CF
            bool isWrapper = typeof(PersistentWrapper).IsAssignableFrom(cls);
#endif
            bool hasTransparentAttribute = cls.GetCustomAttributes(typeof(TransparentPersistenceAttribute), true).Length != 0;

            for (int i = 0; i < flds.Length; i++)
            {
                FieldInfo f = flds[i];
                if (!f.IsNotSerialized && !f.IsStatic)
                {
                    FieldDescriptor fd = new FieldDescriptor();
                    fd.field = f;
                    fd.fieldName = f.Name;
                    fd.className = getTypeName(cls);
                    Type fieldType = f.FieldType;
                    FieldType type = getTypeCode(fieldType);
                    switch (type)
                    {
#if !CF
                        case FieldType.tpInt:
                            if (isWrapper && isObjectProperty(cls, f))
                            {
                                hasReferences = true;
                                type = FieldType.tpOid;
                            }
                            break;
#endif
                        case FieldType.tpArrayOfOid:
                            fd.constructor = GetConstructor(f, "ConstructArray");
                            hasReferences = true;
                            break;
                        case FieldType.tpLink:
                            fd.constructor = GetConstructor(f, "ConstructLink");
                            hasReferences = true;
                            break;

                        case FieldType.tpArrayOfObject:
                        case FieldType.tpObject:
                            hasReferences = true;
                            if (hasTransparentAttribute && isVolanteInternalType(fieldType))
                            {
                                fd.recursiveLoading = true;
                            }
                            break;
                        case FieldType.tpValue:
                            fd.valueDesc = db.getClassDescriptor(f.FieldType);
                            hasReferences |= fd.valueDesc.hasReferences;
                            break;
                        case FieldType.tpArrayOfValue:
                            fd.valueDesc = db.getClassDescriptor(f.FieldType.GetElementType());
                            hasReferences |= fd.valueDesc.hasReferences;
                            break;
                    }
                    fd.type = type;
                    list.Add(fd);
                }
            }
        }

        public static FieldType getTypeCode(System.Type c)
        {
            FieldType type;
            if (c.Equals(typeof(byte)))
            {
                type = FieldType.tpByte;
            }
            else if (c.Equals(typeof(sbyte)))
            {
                type = FieldType.tpSByte;
            }
            else if (c.Equals(typeof(short)))
            {
                type = FieldType.tpShort;
            }
            else if (c.Equals(typeof(ushort)))
            {
                type = FieldType.tpUShort;
            }
            else if (c.Equals(typeof(char)))
            {
                type = FieldType.tpChar;
            }
            else if (c.Equals(typeof(int)))
            {
                type = FieldType.tpInt;
            }
            else if (c.Equals(typeof(uint)))
            {
                type = FieldType.tpUInt;
            }
            else if (c.Equals(typeof(long)))
            {
                type = FieldType.tpLong;
            }
            else if (c.Equals(typeof(ulong)))
            {
                type = FieldType.tpULong;
            }
            else if (c.Equals(typeof(float)))
            {
                type = FieldType.tpFloat;
            }
            else if (c.Equals(typeof(double)))
            {
                type = FieldType.tpDouble;
            }
            else if (c.Equals(typeof(System.String)))
            {
                type = FieldType.tpString;
            }
            else if (c.Equals(typeof(bool)))
            {
                type = FieldType.tpBoolean;
            }
            else if (c.Equals(typeof(System.DateTime)))
            {
                type = FieldType.tpDate;
            }
            else if (c.IsEnum)
            {
                type = FieldType.tpEnum;
            }
            else if (c.Equals(typeof(decimal)))
            {
                type = FieldType.tpDecimal;
            }
            else if (c.Equals(typeof(Guid)))
            {
                type = FieldType.tpGuid;
            }
            else if (typeof(IPersistent).IsAssignableFrom(c))
            {
                type = FieldType.tpObject;
            }
            else if (typeof(ValueType).IsAssignableFrom(c))
            {
                type = FieldType.tpValue;
            }
            else if (typeof(IGenericPArray).IsAssignableFrom(c))
            {
                type = FieldType.tpArrayOfOid;
            }
            else if (typeof(IGenericLink).IsAssignableFrom(c))
            {
                type = FieldType.tpLink;
            }
            else if (c.IsArray)
            {
                type = getTypeCode(c.GetElementType());
                if ((int)type >= (int)FieldType.tpLink)
                {
                    throw new DatabaseException(DatabaseException.ErrorCode.UNSUPPORTED_TYPE, c);
                }
                type = (FieldType)((int)type + (int)FieldType.tpArrayOfBoolean);
            }
            else
            {
                type = FieldType.tpRaw;
            }
            return type;
        }

        internal ClassDescriptor()
        {
        }

        internal ClassDescriptor(DatabaseImpl db, Type cls)
        {
            this.cls = cls;
            name = getTypeName(cls);
            ArrayList list = new ArrayList();
            buildFieldList(db, cls, list);
            allFields = (FieldDescriptor[])list.ToArray(typeof(FieldDescriptor));
            defaultConstructor = cls.GetConstructor(BindingFlags.Instance | BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.DeclaredOnly, null, defaultConstructorProfile, null);
            if (defaultConstructor == null && !typeof(ValueType).IsAssignableFrom(cls))
                throw new DatabaseException(DatabaseException.ErrorCode.DESCRIPTOR_FAILURE, cls);
            resolved = true;
        }

        internal static Type lookup(IDatabase db, String name)
        {
            var resolvedTypes = ((DatabaseImpl)db).resolvedTypes;
            lock (resolvedTypes)
            {
                Type cls;
                var ok = resolvedTypes.TryGetValue(name, out cls);
                if (ok)
                    return cls;
                IClassLoader loader = db.Loader;
                if (loader != null)
                {
                    cls = loader.LoadClass(name);
                    if (cls != null)
                    {
                        resolvedTypes[name] = cls;
                        return cls;
                    }
                }
                Module last = lastModule;
                if (last != null)
                {
                    cls = last.GetType(name);
                    if (cls != null)
                    {
                        resolvedTypes[name] = cls;
                        return cls;
                    }
                }

                int p = name.IndexOf('=');
                if (p >= 0)
                {
                    Type genericType = lookup(db, name.Substring(0, p));
                    Type[] genericParams = new Type[genericType.GetGenericArguments().Length];
                    int nest = 0;
                    int i = p += 2;
                    int n = 0;

                    while (true)
                    {
                        switch (name[i++])
                        {
                            case '[':
                                nest += 1;
                                break;
                            case ']':
                                if (--nest < 0)
                                {
                                    genericParams[n++] = lookup(db, name.Substring(p, i - p - 1));
                                    Debug.Assert(n == genericParams.Length);
                                    cls = genericType.MakeGenericType(genericParams);
                                    if (cls == null)
                                    {
                                        throw new DatabaseException(DatabaseException.ErrorCode.CLASS_NOT_FOUND, name);
                                    }
                                    resolvedTypes[name] = cls;
                                    return cls;
                                }
                                break;
                            case ',':
                                if (nest == 0)
                                {
                                    genericParams[n++] = lookup(db, name.Substring(p, i - p - 1));
                                    p = i;
                                }
                                break;
                        }
                    }
                }

#if CF
                foreach (Assembly ass in DatabaseImpl.assemblies) 
#else
                foreach (Assembly ass in AppDomain.CurrentDomain.GetAssemblies())
#endif
                {
                    foreach (Module mod in ass.GetModules())
                    {
                        Type t = mod.GetType(name);
                        if (t != null)
                        {
                            if (cls != null)
                            {
                                throw new DatabaseException(DatabaseException.ErrorCode.AMBIGUITY_CLASS, name);
                            }
                            else
                            {
                                lastModule = mod;
                                cls = t;
                            }
                        }
                    }
                }
#if !CF
                if (cls == null && name.EndsWith("Wrapper"))
                {
                    Type originalType = lookup(db, name.Substring(0, name.Length - 7));
                    lock (db)
                    {
                        cls = ((DatabaseImpl)db).getWrapper(originalType);
                    }
                }
#endif
                if (cls == null)
                {
                    throw new DatabaseException(DatabaseException.ErrorCode.CLASS_NOT_FOUND, name);
                }
                resolvedTypes[name] = cls;
                return cls;
            }
        }

        public override void OnLoad()
        {
            cls = lookup(Database, name);
            int n = allFields.Length;
            bool hasTransparentAttribute = cls.GetCustomAttributes(typeof(TransparentPersistenceAttribute), true).Length != 0;
            for (int i = n; --i >= 0; )
            {
                FieldDescriptor fd = allFields[i];
                fd.Load();
                fd.field = cls.GetField(fd.fieldName, BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Public);
                if (hasTransparentAttribute && fd.type == FieldType.tpObject && isVolanteInternalType(fd.field.FieldType))
                {
                    fd.recursiveLoading = true;
                }

                switch (fd.type)
                {
                    case FieldType.tpArrayOfOid:
                        fd.constructor = GetConstructor(fd.field, "ConstructArray");
                        break;
                    case FieldType.tpLink:
                        fd.constructor = GetConstructor(fd.field, "ConstructLink");
                        break;
                    default:
                        break;
                }
            }

            defaultConstructor = cls.GetConstructor(BindingFlags.Instance | BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.DeclaredOnly, null, defaultConstructorProfile, null);
            if (defaultConstructor == null && !typeof(ValueType).IsAssignableFrom(cls))
            {
                throw new DatabaseException(DatabaseException.ErrorCode.DESCRIPTOR_FAILURE, cls);
            }
            DatabaseImpl s = (DatabaseImpl)Database;
            if (!s.classDescMap.ContainsKey(cls))
            {
                ((DatabaseImpl)Database).classDescMap.Add(cls, this);
            }
        }

        internal void resolve()
        {
            if (resolved)
                return;

            DatabaseImpl classStorage = (DatabaseImpl)Database;
            ClassDescriptor desc = new ClassDescriptor(classStorage, cls);
            resolved = true;
            if (!desc.equals(this))
            {
                classStorage.registerClassDescriptor(desc);
            }
        }

        public override bool RecursiveLoading()
        {
            return false;
        }
    }
}

File: java/src/org/nachodb/impl/ClassDescriptor.java
package org.nachodb.impl;
import  org.nachodb.*;
import  java.lang.reflect.*;
import  java.util.*;

public final class ClassDescriptor extends Persistent { 
    ClassDescriptor   next;
    String            name;
    boolean           hasReferences;
    FieldDescriptor[] allFields;

    static class FieldDescriptor extends Persistent { 
        String          fieldName;
        String          className;
        int             type;
        ClassDescriptor valueDesc;
        transient Field field;

        public boolean equals(FieldDescriptor fd) { 
            return fieldName.equals(fd.fieldName) 
                && className.equals(fd.className)
                && valueDesc == fd.valueDesc
                && type == fd.type;
        }
    }    

    transient Class       cls;
    transient Constructor loadConstructor;
    transient LoadFactory factory;
    transient Object[]    constructorParams;
    transient boolean     hasSubclasses;
    transient boolean     resolved;
    
    static ReflectionProvider reflectionProvider; 

    public static final int tpBoolean          = 0;
    public static final int tpByte             = 1;
    public static final int tpChar             = 2;
    public static final int tpShort            = 3;
    public static final int tpInt              = 4;
    public static final int tpLong             = 5;
    public static final int tpFloat            = 6;
    public static final int tpDouble           = 7;
    public static final int tpString           = 8;
    public static final int tpDate             = 9;
    public static final int tpObject           = 10;
    public static final int tpValue            = 11;
    public static final int tpRaw              = 12;
    public static final int tpLink             = 13;
    public static final int tpArrayOfBoolean   = 20;
    public static final int tpArrayOfByte      = 21;
    public static final int tpArrayOfChar      = 22;
    public static final int tpArrayOfShort     = 23;
    public static final int tpArrayOfInt       = 24;
    public static final int tpArrayOfLong      = 25;
    public static final int tpArrayOfFloat     = 26;
    public static final int tpArrayOfDouble    = 27;
    public static final int tpArrayOfString    = 28;
    public static final int tpArrayOfDate      = 29;
    public static final int tpArrayOfObject    = 30;
    public static final int tpArrayOfValue     = 31;
    public static final int tpArrayOfRaw       = 32;

    static final String signature[] = {
        "boolean", 
        "byte",
        "char",
        "short",
        "int",
        "long",
        "float",
        "double",
        "String",
        "Date",
        "Object",
        "Value",
        "Raw",
        "Link",
        "", 
        "", 
        "", 
        "", 
        "", 
        "", 
        "ArrayOfBoolean",
        "ArrayOfByte",
        "ArrayOfChar",
        "ArrayOfShort",
        "ArrayOfInt",
        "ArrayOfLong",
        "ArrayOfFloat",
        "ArrayOfDouble",
        "ArrayOfString",
        "ArrayOfDate",
        "ArrayOfObject",
        "ArrayOfValue",
        "ArrayOfRaw"
    };
        

    static final int sizeof[] = {
        1, // tpBoolean
        1, // tpByte
        2, // tpChar
        2, // tpShort
        4, // tpInt
        8, // tpLong
        4, // tpFloat
        8, // tpDouble
        0, // tpString
        8, // tpDate
        4  // tpObject
    };

    static final Class[] perstConstructorProfile = new Class[]{ClassDescriptor.class};

    static ReflectionProvider getReflectionProvider() { 
        if (reflectionProvider == null) { 
            try {
                Class.forName("sun.misc.Unsafe");
                String cls = "org.nachodb.impl.sun14.Sun14ReflectionProvider";
                reflectionProvider = (ReflectionProvider)Class.forName(cls).newInstance();
            } catch (Exception x) { 
                reflectionProvider = new StandardReflectionProvider();
            }
        }
        return reflectionProvider;
    } 
           

    public boolean equals(ClassDescriptor cd) { 
        if (cd == null || allFields.length != cd.allFields.length) { 
            return false;
        }
        for (int i = 0; i < allFields.length; i++) { 
            if (!allFields[i].equals(cd.allFields[i])) { 
                return false;
            }
        }
        return true;
    }
        

    Object newInstance() {
        if (factory != null) { 
            return factory.create(this);
        } else { 
            try { 
                return loadConstructor.newInstance(constructorParams);
            } catch (Exception x) { 
                throw new StorageError(StorageError.CONSTRUCTOR_FAILURE, cls, x);
            }
        }
    }

    void buildFieldList(StorageImpl storage, Class cls, ArrayList list) { 
        Class superclass = cls.getSuperclass();
        if (superclass != null) { 
            buildFieldList(storage, superclass, list);
        }
        Field[] flds = cls.getDeclaredFields();
        for (int i = 0; i < flds.length; i++) { 
            Field f = flds[i];
            if ((f.getModifiers() & (Modifier.TRANSIENT|Modifier.STATIC)) == 0) {
                f.setAccessible(true);
                FieldDescriptor fd = new FieldDescriptor();
                fd.field = f;
                fd.fieldName = f.getName();
                fd.className = cls.getName();
                int type = getTypeCode(f.getType());
                switch (type) {
                  case tpObject:
                  case tpLink:
                  case tpArrayOfObject:
                    hasReferences = true;
                    break;
                  case tpValue:
                    fd.valueDesc = storage.getClassDescriptor(f.getType());
                    hasReferences |= fd.valueDesc.hasReferences;                    
                    break;
                  case tpArrayOfValue:
                    fd.valueDesc = storage.getClassDescriptor(f.getType().getComponentType());
                    hasReferences |= fd.valueDesc.hasReferences;
                }
                fd.type = type;
                list.add(fd);
            }
        }
    }

    public static int getTypeCode(Class c) { 
        int type;
        if (c.equals(byte.class)) { 
            type = tpByte;
        } else if (c.equals(short.class)) {
            type = tpShort;
        } else if (c.equals(char.class)) {
            type = tpChar;
        } else if (c.equals(int.class)) {
            type = tpInt;
        } else if (c.equals(long.class)) {
            type = tpLong;
        } else if (c.equals(float.class)) {
            type = tpFloat;
        } else if (c.equals(double.class)) {
            type = tpDouble;
        } else if (c.equals(String.class)) {
            type = tpString;
        } else if (c.equals(boolean.class)) {
            type = tpBoolean;
        } else if (c.equals(java.util.Date.class)) {
            type = tpDate;
        } else if (IPersistent.class.isAssignableFrom(c)) {
            type = tpObject;
        } else if (IValue.class.isAssignableFrom(c)) {
            type = tpValue;
        } else if (c.equals(Link.class)) {
            type = tpLink;
        } else if (c.isArray()) { 
            type = getTypeCode(c.getComponentType());
            if (type >= tpLink) { 
                throw new StorageError(StorageError.UNSUPPORTED_TYPE, c);
            }
            type += tpArrayOfBoolean;
        } else if (c.equals(Object.class) || c.equals(Comparable.class)) {
            type = tpRaw;
        } else if (serializeNonPersistentObjects) {
            type = tpRaw;            
        } else if (treateAnyNonPersistentObjectAsValue) {
            if (c.equals(Object.class)) { 
                throw new StorageError(StorageError.EMPTY_VALUE);
            }
            type = tpValue;            
        } else { 
            throw new StorageError(StorageError.UNSUPPORTED_TYPE, c);
        }
        return type;
    }

    static boolean treateAnyNonPersistentObjectAsValue = Boolean.getBoolean("perst.implicit.values");
    static boolean serializeNonPersistentObjects = Boolean.getBoolean("perst.serialize.transient.objects");

    ClassDescriptor() {}


    private void locateConstructor() { 
        try { 
            //Class c = Class.forName(cls.getName() + "LoadFactory");
            Class c = Thread.currentThread().getContextClassLoader().loadClass(cls.getName() + "LoadFactory");
            factory = (LoadFactory)c.newInstance();
        } catch (Exception x1) { 
            try {             
                loadConstructor = cls.getDeclaredConstructor(perstConstructorProfile);
                constructorParams = new Object[]{this};
            } catch (NoSuchMethodException x2) {
                try { 
                    loadConstructor = getReflectionProvider().getDefaultConstructor(cls);
                    constructorParams = null;
                } catch (Exception x3) {
                    throw new StorageError(StorageError.DESCRIPTOR_FAILURE, cls, x3);
                }
            }
            loadConstructor.setAccessible(true);
        }
    }

    ClassDescriptor(StorageImpl storage, Class cls) { 
        this.cls = cls;
        name = cls.getName();
        ArrayList list = new ArrayList();
        buildFieldList(storage, cls, list);
        allFields = (FieldDescriptor[])list.toArray(new FieldDescriptor[list.size()]);
        locateConstructor();
        resolved = true;
    }

    protected static Class loadClass(Storage storage, String name) { 
        ClassLoader loader = storage.getClassLoader();
        if (loader != null) { 
            try { 
                return loader.loadClass(name);
            } catch (ClassNotFoundException x) {}
        }
        try { 
            return Thread.currentThread().getContextClassLoader().loadClass(name);
            // return Class.forName(name);
        } catch (ClassNotFoundException x) { 
            throw new StorageError(StorageError.CLASS_NOT_FOUND, name, x);
        }
    }

    public void onLoad() {         
        cls = loadClass(getStorage(), name);
        Class scope = cls;
        int n = allFields.length;
        for (int i = n; --i >= 0;) { 
            FieldDescriptor fd = allFields[i];
            fd.load();
            if (!fd.className.equals(scope.getName())) {
                for (scope = cls; scope != null; scope = scope.getSuperclass()) { 
                    if (fd.className.equals(scope.getName())) {
                        break;
                    }
                }
            }
            if (scope != null) {
                try { 
                    Field f = scope.getDeclaredField(fd.fieldName);
                    if ((f.getModifiers() & (Modifier.TRANSIENT|Modifier.STATIC)) == 0) {
                        f.setAccessible(true);
                        fd.field = f;
                    }
                } catch (NoSuchFieldException x) {}
            } else { 
                scope = cls;
            }
        }
        for (int i = n; --i >= 0;) { 
            FieldDescriptor fd = allFields[i];
            if (fd.field == null) { 
            hierarchyLoop:
                for (scope = cls; scope != null; scope = scope.getSuperclass()) { 
                    try { 
                        Field f = scope.getDeclaredField(fd.fieldName);
                        if ((f.getModifiers() & (Modifier.TRANSIENT|Modifier.STATIC)) == 0) {
                            for (int j = 0; j < n; j++) { 
                                if (allFields[j].field == f) { 
                                    continue hierarchyLoop;
                                }
                            }
                            f.setAccessible(true);
                            fd.field = f;
                            break;
                        }
                    } catch (NoSuchFieldException x) {}
                }
            }
        }
        locateConstructor();
        StorageImpl s = (StorageImpl)getStorage();
        if (s.classDescMap.get(cls) == null) { 
            s.classDescMap.put(cls, this);
        }
    }

       
    void resolve() {
        if (!resolved) { 
            StorageImpl classStorage = (StorageImpl)getStorage();
            ClassDescriptor desc = new ClassDescriptor(classStorage, cls);
            resolved = true;
            if (!desc.equals(this)) { 
                classStorage.registerClassDescriptor(desc);
            }
        }
    }            

    public boolean recursiveLoading() { 
        return false;
    }
}
